### PR TITLE
[generate_vmap_rule] add restore_vmap helper function

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91211
* #91206
* #90966
* #90965
* #90964
* __->__ #90963

As seen in
https://docs.google.com/document/d/1bIQkWXy3J35_20c_a5kchikabBW5M8_uRAhl0BIMwU4/edit

`restore_vmap` is a private helper function. It is vmap but has the
following
differences:
- instead of returning outputs, it returns an (outputs, out_dims) tuple.
  out_dims is a pytree of shape shape as outputs and contains Optional[int]
  specifying where the vmapped dimension, if it exists, is in the
  corresponding output.
- does no validation on in_dims or inputs (vmap expects at least one
  Tensor to be vmapped).
  restore_vmap allows for no inputs to have the vmap dimension
- does no validation on outputs (vmap expects only Tensor outputs)
  restore_vmap allows for return of arbitrary outputs (not just
  Tensors)

Test Plan:
- added some simple test to test restore_vmap
- I am OK with restore_vmap not being a part of vmap right now -- the
implementation of vmap rarely changes and it is a bit difficult to
refactor vmap in a way that restore_vmap is a subroutine.

Other questions:
- Bikeshedding the `restore_vmap` name

### CODE REVIEW

**Review by samdow — APPROVED** [review]
re: bikeshedding, I think `restore_vmap` makes sense with how it's used later on. Might be nice to add a note about this to the comment but it's already a long comment so not attached

File: torch/_functorch/vmap.py [hunk 1]
```diff
@@ -490,3 +490,52 @@ def _flat_vmap(func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, r
...
+
+# `restore_vmap` is a private helper function. It is vmap but has the following
+# differences:
+# - instead of returning outputs, it returns an (outputs, out_dims) tuple.
+#   out_dims is a pytree of shape shape as outputs and contains Optional[int]
+#   specifying where the vmapped dimension, if it exists, is in the corresponding output.
+# - does no validation on in_dims or inputs (vmap expects at least one Tensor to be vmapped).
+#   restore_vmap allows for no inputs to have the vmap dimension
+# - does no validation on outputs (vmap expects only Tensor outputs)
+#   restore_vmap allows for return of arbitrary outputs (not just Tensors)
```
Comment (soulitzer) [inline torch/_functorch/vmap.py:503]: Might be good to mention why these relaxations are needed.

**Review by soulitzer — APPROVED** [review]

**Comment by soulitzer** [comment]
I also agree that `restore_vmap` makes sense for how it used, but if we wanted to name it based on what it does, we could do`vmap_no_check`? Naming it based on what it does might be better if we plan to use it later for other purposes.

**Comment by zou3519** [comment]
> I also agree that `restore_vmap` makes sense for how it used, but if we wanted to name it based on what it does, we could do something in the spirit of `vmap_no_check`? Naming it based on what it does might be better if we plan to use it later for other purposes.

I'll keep it as restore_vmap for now since we think it makes sense for how it's used. We can rename it in the future if it gets used for other purposes.

**Pull request closed without merging**